src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,602 @@
+open Core_kernel
+open Middle
+open Middle.Expr
+
+(**
+ * Recursivly look in Decls for sized types that hold matrices or vectors
+ *)
+let get_eigen_decls Stmt.Fixed.({pattern; _}) =
+  match pattern with
+  | Stmt.Fixed.Pattern.Decl
+      { decl_adtype= UnsizedType.AutoDiffable
+      ; decl_id
+      ; decl_type= Type.Sized sized_type }
+    when SizedType.contains_eigen_type sized_type ->
+      Set.Poly.singleton decl_id
+  | _ -> Set.Poly.empty
+
+(**
+ * Modify an index's inner expressions with `op`
+ * @param op a functor returning an expression
+ * @param ind the Index.t to modify
+ *)
+let modify_index op ind =
+  match ind with
+  | Index.All -> Index.All
+  | Single ind_expr -> Single (op ind_expr)
+  | Upfrom ind_expr -> Upfrom (op ind_expr)
+  | Between (expr_top, expr_bottom) -> Between (op expr_top, op expr_bottom)
+  | MultiIndex exprs -> MultiIndex (op exprs)
+
+(**
+ * Apply an op returning a Set of strings for `Index` types
+ *  with inner expressions
+ * @param op a functor returning a Set of strings
+ * @param ind the Index.t to modify
+ *)
+let map_index op ind =
+  match ind with
+  | Index.All -> Set.Poly.empty
+  | Single ind_expr -> op ind_expr
+  | Upfrom ind_expr -> op ind_expr
+  | Between (expr_top, expr_bottom) ->
+      Set.Poly.union (op expr_top) (op expr_bottom)
+  | MultiIndex exprs -> op exprs
+
+(**
+ * Apply a functor to an `Index` type's inner returning a boolean if 
+ * the conditional is true for any of the inner expressions.
+ *)
+let name_exists_in_index op ind =
+  match ind with
+  | Index.All -> false
+  | Single ind_expr -> op ind_expr
+  | Upfrom ind_expr -> op ind_expr
+  | Between (expr_top, expr_bottom) -> op expr_top || op expr_bottom
+  | MultiIndex exprs -> op exprs
+
+(**
+ * Search through an expression for `Var name` where `name` is in
+ *  the `aos_exits`
+ * and the type of the expression contains an eigen type.
+ * @param aos_exits A Set of strings with the name of the obj
+ *  we are searching for.
+ * @param pattern An expression pattern we recursivly search through 
+ *  for the name.
+ **)
+let rec query_expr_eigen_names (aos_exits : string Set.Poly.t)
```

Comment:
Can you do `not (Set.Poly.is_empty (Set.inter aos_exits (query_all_expr_eigen_names expr)))`? To avoid rewriting the traversal logic

Reply:
Yep!